### DISCUSSION TITLE
pandas with custom extension arrays cannot be queried

### DISCUSSION THREAD

**Discussion by achapkowski** [duckdb/duckdb#5094] [General]
### What happens?

A pandas' dataframe with a custom accessor and dtype fail to query with the following message:

```
results = con.execute("SELECT * FROM sdf")
```

Results in 

```
NotImplementedException: Not implemented Error: Data type 'geometry' not recognized
```

### To Reproduce

1. Create an extension array and dtype as defined here: https://pandas.pydata.org/pandas-docs/stable/development/extending.html?highlight=extensionarray
2. Create a pandas dataframe with the custom data type
3. run: results = con.execute("SELECT * FROM sdf")
4. Failure

### OS:

windows 10 and 11

### DuckDB Version:

0.5.1

### DuckDB Client:

python

### Full Name:

andrew

### Affiliation:

don't know why this matters

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by hannes** [comment]
This is indeed not supported. Maybe @maxxen wants to consider it as part as the overall Geo effort. As a note, stating your affiliation matters, but in your case it's clear from your profile.

**Reply by achapkowski** [reply]
asking for full names and affiliations seems a bit intrusive TBH. 

That isn't really the issue here though, taking the API for the spin, I think you need to support this because custom dtypes is part of the pandas API.

**Reply by achapkowski** [reply]
For example, sqlite doesn't inherently support custom data types, but it has a register method that allows users to do this.  Could the same thing be implemented?  

``` sqlite3.register_adapter``` see: https://docs.python.org/3/library/sqlite3.html#how-to-register-adapter-callables

**Comment by Maxxen** [comment]
FWIW, interoperability with the corresponding geo extensions of the systems we currently integrate with, like geopandas for pandas, is definitely something I have in mind, along with geoarrow/geoparquet once they mature. But Im not even close to be able to provide a time frame, It's going to be implemented post geo-extension MVP at the very least.

**Reply by achapkowski** [reply]
I wasn't asking really about Geo more the ability to have the duckdb understand nonstandard types.

**Reply by Alex-Monahan** [reply]
Looks like they just became non-experimental with this version of Pandas!
Note

The pandas.api.extensions.ExtensionDtype and pandas.api.extensions.ExtensionArray APIs were experimental prior to pandas 1.5. Starting with version 1.5, future changes will follow the pandas deprecation policy.

https://pandas.pydata.org/pandas-docs/stable/development/extending.html?highlight=extensionarray

**Reply by pankajp** [reply]
While the general ExtensionArray api seems not very suitable for integration with duckdb (python element extraction would be a lot of overhead and just calling methods on the extension arrays might not be featured enough to implement full sql, and definitely not performant)
What duckdb could do is to handle arrow convertible extension types:
https://pandas.pydata.org/pandas-docs/stable/development/extending.html?highlight=extensionarray#compatibility-with-apache-arrow
That would already support a lot of pandas extension types (nullable ints and strings) and I believe pyarrow integration already exists in duckdb so should be much less code to add.

**Reply by achapkowski** [reply]
My custom extension implements this already and it doesn't work.  The arrow type is binary representation.
